h/search/query.py

```diff
@@ -242,6 +286,95 @@ def __call__(self, search, params):
+        Values containing wildcards are intrepted as wildcard searches.
+
+    If specifying a uri with wildcards:
+    * will match any character sequence (including an empty one), and a ? will match
+    any single character.
+    """
+
+    def __init__(self, request, separate_keys=False):
+        """Initialize a new UriFilter.
+
+        :param request: the pyramid.request object
+        :param separate_keys: if True will treate wildcard_uri as wildcards and uri/url
+            as exact match. If False will treat any values in uri/url containing wildcards
+            ("?" or "*") as wildcard searches.
+
+        """
+        self.request = request
+        self.separate_keys = separate_keys
+
+    def __call__(self, search, params):
+        # If there are no uris to search for there is nothing to do so return early.
+        if not ("uri" in params or "url" in params or "wildcard_uri" in params):
+            return search
+
+        if self.separate_keys:
+            uris = _pop_param_values(params, 'uri') + _pop_param_values(params, 'url')
+            wildcard_uris = _pop_param_values(params, 'wildcard_uri')
+        else:
+            uris = _pop_param_values(params, 'uri') + _pop_param_values(params, 'url')
+            # Split into wildcard uris and non wildcard uris.
+            wildcard_uris = [u for u in uris if "*" in u or "?" in u]
+            uris = [u for u in uris if "*" not in u and "?" not in u]
+
+        # Ignore all invalid wildcard uris.
+        wildcard_uris = self._normalize_uris(
+            [u for u in wildcard_uris if wildcard_uri_is_valid(u)],
+            normalize_method=self._wildcard_uri_normalized)
+        uris = self._normalize_uris(uris)
+
+        queries = []
+        if wildcard_uris:
+            queries = [Q('wildcard', **{"target.scope": u}) for u in wildcard_uris]
+        if uris:
+            queries.append(Q("terms", **{'target.scope': uris}))
+        return search.query('bool', should=queries)
+
+    def _normalize_uris(self, query_uris, normalize_method=uri.normalize):
+        uris = set()
+        for query_uri in query_uris:
+            expanded = storage.expand_uri(self.request.db, query_uri)
+
+            us = [normalize_method(u) for u in expanded]
+            uris.update(us)
+        return list(uris)
+
+    def _wildcard_uri_normalized(self, wildcard_uri):
+        """
+        Same as uri.normalized but it doesn't strip ending `?` from uri's.
+
+        It's possible to have a wildcard at the end of a uri, however
+        uri.normalize strips `?`s from the end of uris and something like
+        http://foo.com/* will not be normalized to http://foo.com* without
+        removing the `*` before normalization. To compinsate for this,
+        we check for an ending wildcard and add it back after normalization.
+
+        While it's possible to escape `?` and `*` using `\`, the uri.normalize
```

Comment:
If you use `\` in a string literal (regular string literal or docstring) and are not intending to escape a character, I think this needs to be a raw string - following one of @seanh's most recent PRs.